ui/src/main/java/edu/wpi/grip/ui/MainWindowController.java

```diff
@@ -275,33 +286,64 @@ protected void quit() {
-    fileChooser.getExtensionFilters().add(new ExtensionFilter(Language.CPP.name, "*.cpp"));
-    fileChooser.getExtensionFilters().add(new ExtensionFilter(Language.PYTHON.name, "*.py"));
-    fileChooser.setInitialFileName("Pipeline.java");
-    final File file = fileChooser.showSaveDialog(root.getScene().getWindow());
-    if (file == null) {
+  @FXML
+  protected void generate() {
+    if (pipeline.getSources().isEmpty()) {
+      // No sources
+      // TODO show warning alert (#693)
       return;
-    }
-    Language lang = Language.get(fileChooser.getSelectedExtensionFilter().getDescription());
-    Exporter exporter = new Exporter(pipeline.getSteps(), lang, file);
-    final Set<String> nonExportableSteps = exporter.getNonExportableSteps();
-    if (!nonExportableSteps.isEmpty()) {
-      StringBuilder b = new StringBuilder("The following steps cannot be exported:\n");
-      nonExportableSteps.forEach(n -> b.append("  ").append(n).append('\n'));
-      Alert alert = new Alert(Alert.AlertType.WARNING);
-      alert.setContentText(b.toString());
-      alert.showAndWait();
+    } else if (pipeline.getSteps().isEmpty()) {
+      // Sources, but no steps
+      // TODO show warning alert (#693)
+      return;
+    } else if (pipeline.getConnections().isEmpty()) {
+      // Sources and steps, but no connections
+      // TODO show warning alert (#693)
+      return;
+    } else if (pipeline.getSteps().stream()
+        .flatMap(s -> s.getInputSockets().stream())
+        .filter(s -> SocketHint.View.NONE.equals(s.getSocketHint().getView()))
+        .anyMatch(s -> !s.getValue().isPresent())) {
+      // Some sockets aren't connected
+      // TODO show warning alert (#693)
       return;
     }
-    Thread exportRunner = new Thread(exporter);
-    exportRunner.setDaemon(true);
-    exportRunner.start();
+    Dialog<CodeGenerationOptions> optionsDialog = new ChoiceDialog<>();
+    optionsDialog.setTitle("Code Generation Options");
+    optionsDialog.setHeaderText(null);
+    optionsDialog.setGraphic(null);
+    optionsDialog.getDialogPane().setContent(codegenOptionsPane);
+    CodeGenerationOptionsController c =
+        (CodeGenerationOptionsController) codegenOptionsPane.getProperties().get("controller");
+    optionsDialog.setResultConverter(bt -> bt.getButtonData() == null ? null
+        : bt.getButtonData() == ButtonBar.ButtonData.OK_DONE
+        ? c.getOptions() : null);
+    Optional<CodeGenerationOptions> o = optionsDialog.showAndWait();
+    if (o.isPresent()) {
+      CodeGenerationOptions options = o.get();
+      ProjectSettings s = settingsProvider.getProjectSettings();
+      s.setCodegenDestDir(new File(options.getSaveDir()));
+      s.setGeneratedJavaPackage(options.getPackageName());
+      s.setGeneratedPipelineName(options.getClassName());
+      s.setGeneratedPythonModuleName(options.getModuleName());
+      s.setPreferredGeneratedLanguage(options.getLanguage().name);
```

Comment:
Should probably have a `CodeGenerationOptions.copyTo(ProjectSettings)` method to avoid this crap